### PR TITLE
Player 1532

### PR DESCRIPTION
**Pull request by dallen42**
PR for updating iq.js to handle QoS events

### CODE REVIEW

**Comment by pilievOoyala** [comment]
LGTM.  @aeng7 can you review too.

File: js/framework/AnalyticsConstants.js [hunk 1]
```diff
@@ -885,6 +959,190 @@ if (!OO.Analytics.EVENT_DATA)
...
+   * @classdesc Contains the information about the initial play starting event.
+   * @property {string} playerCoreVersion The player core version
+   * @property {number} timeSinceInitialPlay The time since the initial play request was made
+   * @property {boolean} autoplayed Boolean for if the video was autoplayed or not
+   * @property {boolean} hadPreroll Boolean for if the video had an ad play before it started
+   * @property {number} position The initial position of the playhead upon playback start. This includes 
+   *   midrolls that play before content due to an initial playhead time > 0
+   * @property {string} plugin The video plugin used for playback
+   * @property {string} technology The browser technology used - HTML5, Flash, Mixed, or Other
+   * @property {string} encoding The stream encoding type, i.e. MP4, HLS, Dash, etc.
```
Comment (aeng7) [inline js/framework/AnalyticsConstants.js]: Should this be a string or object? Code below checks if it is an object:

this.encoding = checkInitialPlayStartingData(encoding, "encoding", ["object"]);

File: js/framework/AnalyticsConstants.js [hunk 2]
```diff
@@ -885,6 +959,190 @@ if (!OO.Analytics.EVENT_DATA)
...
+    this.playerCoreVersion = checkPlaybackReadyData(playerCoreVersion, "playerCoreVersion", ["string"]);
+    this.timeSincePlayerCreated = checkPlaybackReadyData(timeSincePlayerCreated, "timeSincePlayerCreated", ["number"]);
+    this.pluginList = checkPlaybackReadyData(pluginList, "pluginList", ["array"]);
+  };
+
+  /**
+   * @public
+   * @class Analytics.EVENT_DATA#ApiErrorData
+   * @classdesc Contains information about the api error.
+   * @property {object} errorCode The error code if any
```
Comment (aeng7) [inline js/framework/AnalyticsConstants.js]: Should this be an object or number? Code below checks if this is a number:

this.errorCode = checkApiErrorData(errorCode, "errorCode", ["number"]);

File: js/plugins/iq.js [hunk 3]
```diff
@@ -160,11 +160,21 @@ var IqPlugin= function (framework)
       case OO.Analytics.EVENTS.VIDEO_PLAYER_CREATED:
         if (params && params[0])
         {
-          pcode = params[0].pcode;
-          playerId = params[0].playerBrandingId;
+          eventParams = params[0];
+          pcode = eventParams.params.pcode;
```
Comment (aeng7) [inline js/plugins/iq.js]: Can we add a null check to see if eventParams.params exists?

File: js/plugins/iq.js [hunk 4]
```diff
@@ -236,7 +246,29 @@ var IqPlugin= function (framework)
         this.ooyalaReporter.reportReplay();
         OO.log("IQ: Reported: reportReplay()");
         break;
-
+      //OO.EVENTS.BUFFERING -> OO.Analytics.EVENTS.VIDEO_BUFFERING_STARTED.
+      case OO.Analytics.EVENTS.VIDEO_BUFFERING_STARTED: 
+        eventParams = params[0];
```
Comment (aeng7) [inline js/plugins/iq.js]: Can we add a null check to see if params and params[0] exists?

File: js/plugins/iq.js [hunk 5]
```diff
@@ -236,7 +246,29 @@ var IqPlugin= function (framework)
...
+      case OO.Analytics.EVENTS.INITIAL_PLAY_STARTING:
+      case OO.Analytics.EVENTS.PLAYBACK_READY:
+      case OO.Analytics.EVENTS.API_ERROR:
+      case OO.Analytics.EVENTS.BITRATE_INITIAL:
+      case OO.Analytics.EVENTS.BITRATE_FIVE_SEC:
+      case OO.Analytics.EVENTS.BITRATE_STABLE:
+      case OO.Analytics.EVENTS.PLAYBACK_START_ERROR:
+      case OO.Analytics.EVENTS.PLAYBACK_MIDSTREAM_ERROR:
+      case OO.Analytics.EVENTS.PLUGIN_LOADED:
+        eventMetadata = params[0];
```
Comment (aeng7) [inline js/plugins/iq.js]: Can we add a null check to see if params and params[0] exists?

File: test/unit-test-helpers/AnalyticsFrameworkTestUtils.js [hunk 6]
```diff
@@ -211,9 +211,14 @@ if (!OO.Analytics.Utils)
...
-          pcode: metadata.pcode,
-          playerBrandingId: metadata.playerBrandingId
+        plugin.processEvent(OO.Analytics.EVENTS.VIDEO_PLAYER_CREATED, [{ 
+          params:{
+            pcode: metadata.pcode,
+            playerBrandingId: metadata.playerBrandingId
+          },
+          embedCode: metadata.embedCode,
+          playerUrl: metadata.playerUrl,
+          playerCoreVersion : "v4"
```
Comment (aeng7) [inline test/unit-test-helpers/AnalyticsFrameworkTestUtils.js]: Do we need to update the VIDEO_PLAYER_CREATED documentation with these new params?

File: js/plugins/iq.js [hunk 7]
```diff
@@ -236,7 +246,29 @@ var IqPlugin= function (framework)
...
+        eventMetadata.qosEventName = eventName;
+        OO.log("IQ: Reported: reportCustomEvent() for event: " + eventName + " with args:" + JSON.stringify(eventMetadata));
+        this.ooyalaReporter.reportCustomEvent(eventName, eventMetadata);
+        break;
```
Comment (aeng7) [inline js/plugins/iq.js]: Can we add unit tests for these new event handlers?

File: test/unit-test-helpers/AnalyticsFrameworkTestUtils.js (hunk 6 above)
Comment (dallen42) [reply test/unit-test-helpers/AnalyticsFrameworkTestUtils.js]: These aren't actually extra parameters, they are properties defined in the object. The only parameter will still be a single array of objects, as with the other events. Most of those events are similarly defined, with "paramArray Array of length 1, containing the original parameters" being the only parameter described in the docs. None any member attributes of the objects within the array are mentioned

File: js/plugins/iq.js (hunk 7 above)
Comment (dallen42) [reply js/plugins/iq.js]: Do you mean something to test reportCustomEvent(), or do you want a unit test for each event checked by the switch case statement?

File: js/framework/AnalyticsConstants.js (hunk 1 above)
Comment (dallen42) [reply js/framework/AnalyticsConstants.js]: Should be string, fixed

File: js/framework/AnalyticsConstants.js (hunk 2 above)
Comment (dallen42) [reply js/framework/AnalyticsConstants.js]: Should be a number, fixed

File: js/plugins/iq.js (hunk 3 above)
Comment (dallen42) [reply js/plugins/iq.js]: I added it as a check on params[0].params, since params[0] == eventParams

File: js/plugins/iq.js (hunk 4 above)
Comment (dallen42) [reply js/plugins/iq.js]: Added

File: js/plugins/iq.js (hunk 5 above)
Comment (dallen42) [reply js/plugins/iq.js]: Added

File: js/framework/AnalyticsConstants.js [hunk 8]
```diff
@@ -885,6 +959,190 @@ if (!OO.Analytics.EVENT_DATA)
...
+
+  /**
+   * @public
+   * @class Analytics.EVENT_DATA#VideoPlayerCreatedData
+   * @classdesc Contains information about the player created event
+   * @property {string} playerCoreVersion The player core version
+   * @property {object} params The configuration metadata associated with the player
+   * (i.e. pcode, playerBrandingId, skin configuration, player configuration parameters)
+   * @property {string} embedCode The embed code of the asset attempting to play
+   * @property {string} playerUrl The url of the page containing the player
```
Comment (pilievOoyala) [inline js/framework/AnalyticsConstants.js]: pcode is missing from the jsdoc

Comment (dallen42) [reply js/framework/AnalyticsConstants.js]: Fixed

File: js/plugins/iq.js (hunk 7 above)
Comment (aeng7) [reply js/plugins/iq.js]: Each event checked by the switch statement.

Comment (dallen42) [reply js/plugins/iq.js]: @aeng7 added a new test for each of the qos events, working on ad events now

Comment (aeng7) [reply js/plugins/iq.js]: Awesome, thanks!

Comment (dallen42) [reply js/plugins/iq.js]: @aeng7 Added a bunch of unit tests that were missing for the ad events, should be good for unit tests now

File: test/unit-tests/testIq.js [hunk 9]
```diff
@@ -370,14 +377,592 @@ describe('Analytics Framework Template Unit Tests', function()
...
+    simulator.simulateVideoBufferingStarted({position : testPosition});
+
+    var unitTestState = plugin.ooyalaReporter.unitTestState;
+    expect(unitTestState.reportCustomEventCalled).toBe(1);
+    expect(unitTestState.eventName).toBe(eventName);
+    expect(unitTestState.eventMetadata.qosEventName).toBe(eventName);
+    expect(unitTestState.eventMetadata.position).toBe(testPosition);
+  });
+
+  // TODO: send actual params and check eventMetadata to ensure they are sent
```
Comment (aeng7) [inline test/unit-tests/testIq.js]: Is this TODO still valid? It seems like it is being done in the test below.

**Pull request merged**